File: frontend/src/types/budget.ts
export interface BudgetCategory {
  category: string  // Category ID (e.g., "FOOD_OUT" or custom)
  spending: number
  cap: number
  percentage: number
  remaining: number
  icon?: string // Lucide icon name
  color?: string // Hex color
  display_name?: string // Display name from custom category
}

export interface BudgetStatus {
  year: number
  month: number
  month_name: string
  categories: BudgetCategory[]
  total_spending: number
  total_cap: number
  total_percentage: number
  total_remaining: number
  excluded_categories: string[]  // Category IDs excluded from total calculation
  period_start: string      // ISO date
  period_end: string        // ISO date
  period_label: string      // "Mar 15 – Apr 14, 2026"
  days_in_period: number
  days_elapsed: number
  monthly_total_cap: number // Original monthly cap for reference
}

export interface BudgetCapsUpdate {
  total_budget: number
  category_budgets: Record<string, number>
}

export interface BudgetCapsResponse {
  success: boolean
  message: string
  updated_caps: Record<string, number>
}

File: ios/BudgetMaster/Models/Budget.swift
import Foundation

// MARK: - BudgetCategory

public struct BudgetCategory: Codable, Sendable, Identifiable {
    public var id: String { category }

    public let category: String
    public let spending: Double
    public let cap: Double
    public let percentage: Double
    public let remaining: Double
    public let emoji: String
}

// MARK: - BudgetStatus

public struct BudgetStatus: Codable, Sendable {
    public let year: Int
    public let month: Int
    public let monthName: String
    public let categories: [BudgetCategory]
    public let totalSpending: Double
    public let totalCap: Double
    public let totalPercentage: Double
    public let totalRemaining: Double
    public let excludedCategories: [String]
}

// MARK: - BulkBudgetUpdateRequest

public struct BulkBudgetUpdateRequest: Codable, Sendable {
    public let totalBudget: Double
    public let categoryBudgets: [String: Double]

    public init(totalBudget: Double, categoryBudgets: [String: Double]) {
        self.totalBudget = totalBudget
        self.categoryBudgets = categoryBudgets
    }
}

// MARK: - BulkBudgetUpdateResponse

public struct BulkBudgetUpdateResponse: Codable, Sendable {
    public let success: Bool
    public let message: String
    public let updatedCaps: [String: Double]
}

// MARK: - TotalBudgetResponse

public struct TotalBudgetResponse: Codable, Sendable {
    public let totalMonthlyBudget: Double
    public let allocated: Double
    public let available: Double
}

// MARK: - TotalBudgetUpdateRequest

public struct TotalBudgetUpdateRequest: Codable, Sendable {
    public let totalMonthlyBudget: Double

    public init(totalMonthlyBudget: Double) {
        self.totalMonthlyBudget = totalMonthlyBudget
    }
}

// MARK: - TotalBudgetUpdateResponse

public struct TotalBudgetUpdateResponse: Codable, Sendable {
    public let success: Bool
    public let totalMonthlyBudget: Double
    public let otherCap: Double
    public let message: String
}
